### PR TITLE
Mongo: provide a mechanism to delay mongo calls.

### PR DESCRIPTION
**Pull request by RomanDzhabarov**
This PR introduces injection of delays into mongo requests.

### CODE REVIEW

File: source/common/mongo/proxy.cc [hunk 1]
```diff
@@ -241,5 +245,54 @@ DecoderPtr ProdProxyFilter::createDecoder(DecoderCallbacks& callbacks) {
...
+                                          fault_config_->delayPercent())) {
+    return result;
+  }
+
+  uint64_t duration = runtime_.snapshot().getInteger("mongo.fault.delay.duration_ms",
+                                                     fault_config_->delayDuration());
+
+  // Delay only if the duration is > 0ms.
+  if (duration > 0) {
+    result.value(duration);
```
Comment (htuch) [inline source/common/mongo/proxy.cc]: Drive-by-question: would it make sense to pick this from a distribution bounded by duration, or do you want exactly the same delay each time?

Comment (RomanDzhabarov) [inline source/common/mongo/proxy.cc]: It makes sense to have faults which are distribution driven as well.
This PR is for the simple case to get a parity with HTTP faults filter and kicking off initial testing. Current implementation injecting fixed latency for some % of requests.
I can name it explicitly fixed_delay to avoid confusions as well.

**Comment by mattklein123** [comment]
I won't be able to review till tomorrow. @ccaraman can you take a first pass?

@RomanDzhabarov please just add docs in this PR while we are going through reviews. It's not that much to add and there is no reason to not do them now. Thanks.

File: docs/configuration/network_filters/mongo_proxy_filter.rst [hunk 2]
```diff
@@ -25,6 +26,30 @@ access_log
   path is specified no access logs will be written. Note that access log is also gated by
   :ref:`runtime <config_network_filters_mongo_proxy_runtime>`.
 
+fault
+  *(optional, object)* If specified, the filter will inject faults  based on the values in the object.
```
Comment (ccaraman) [inline docs/configuration/network_filters/mongo_proxy_filter.rst]: nit: extra space between faults and based

**Comment by RomanDzhabarov** [comment]
The last job failed for not PR specific reasons. Retriggering it.

File: source/common/mongo/proxy.h [hunk 3]
```diff
@@ -73,6 +77,31 @@ class AccessLog {
...
+
+class FaultConfigImpl : public FaultConfig {
+public:
+  FaultConfigImpl(const Json::Object& fault_config);
+
+  uint32_t delayPercent() const override { return delay_percent_; }
+  uint64_t delayDuration() const override { return duration_ms_; }
+
+private:
+  uint32_t delay_percent_;
```
Comment (ccaraman) [inline source/common/mongo/proxy.h]: make these const

Comment (ccaraman) [inline source/common/mongo/proxy.h]: same for duration_ms_

Comment (RomanDzhabarov) [inline source/common/mongo/proxy.h]: if I do `const` here then delay_percent_ has to be in the initializer list something like

FaultConfigImpl::FaultConfigImpl(const Json::Object& fault_config) : 
   delay_percent_(static_cast<uint32_t>(fault_config.getObject("fault")->getInteger("percent"))),
 ...

Let me know, i'm not thrilled with the construction above but could do it either way.

File: source/common/mongo/proxy.h [hunk 4]
```diff
@@ -73,6 +77,31 @@ class AccessLog {
 typedef std::shared_ptr<AccessLog> AccessLogSharedPtr;
 
 /**
+ * Mongo fault configuration.
+ */
+class FaultConfig {
```
Comment (ccaraman) [inline source/common/mongo/proxy.h]: no need to have an interface.

Comment (RomanDzhabarov) [inline source/common/mongo/proxy.h]: this is to abstract the way FaultConfig is constructed, so to have a handy (simplistic) implementation for the test.
This provides additional flexibility for testing purposes.

File: source/common/mongo/proxy.h (hunk 3 above)
Comment (RomanDzhabarov) [inline source/common/mongo/proxy.h]: this will be fixed and talked offline

File: source/common/mongo/proxy.cc [hunk 5]
```diff
@@ -241,5 +245,54 @@ DecoderPtr ProdProxyFilter::createDecoder(DecoderCallbacks& callbacks) {
...
+  }
+
+  return result;
+}
+
+void ProxyFilter::delayInjectionTimerCallback() {
```
Comment (ccaraman) [inline source/common/mongo/proxy.cc]: this isn't the callback that you want. We want a delay to operate on an active query. When the timer trigggers this code is telling the filter to continue on reading instead of processing the activequery that the delay should be associated it.

Comment (RomanDzhabarov) [inline source/common/mongo/proxy.cc]: Active query list is used just to match request/response in order to capture timing and other full request processing stats.

When new operation coming in `Query` for example, we recognize that it's a `Query` operation and we delay all of the further processing (if needed). The active query will be processed once we get a reply for that (based on id).

I do not see any better way of handling this but open for any suggestions.

File: source/common/mongo/proxy.h (hunk 4 above)
Comment (ccaraman) [inline source/common/mongo/proxy.h]: No there is a way to write the tests without having to create a TestFaultConfig, see my comment below.

File: test/common/mongo/proxy_test.cc [hunk 6]
```diff
@@ -64,23 +76,89 @@ class MongoProxyFilterTest : public testing::Test {
...
+
+  void initializeFilter() {
+    filter_.reset(
+        new TestProxyFilter("test.", store_, runtime_, access_log_, fault_config_, dispatcher_));
     filter_->initializeReadFilterCallbacks(read_filter_callbacks_);
     filter_->onNewConnection();
   }
 
+  void setupDelayFault(uint64_t duration_ms, uint32_t delay_percent, bool enable_fault) {
+    TestFaultConfig* test_config = new TestFaultConfig();
```
Comment (ccaraman) [inline test/common/mongo/proxy_test.cc]: Create a real FaultConfig here and just pass in a JSON.

File: source/common/mongo/proxy.cc (hunk 5 above)
Comment (mattklein123) [inline source/common/mongo/proxy.cc]: What you have now is not correct. Mongo can be multiplexing, so you can get 2 queries in a decode for example. If the 2nd query continues while there is an outstanding delay operation, badness will occur. You can potentially handle like you are doing, but you need to consider this case. (There are other problems with this PR BTW but will let @ccaraman continue with review).

File: source/common/mongo/proxy.h (hunk 4 above)
Comment (RomanDzhabarov) [inline source/common/mongo/proxy.h]: That's correct, I did not say there is no way to test it.

This is to avoid relying on json for testing filter functionality, which inherently introduces additional code complexity.
Going to stick to less code == fewer bugs for now.

File: source/common/mongo/proxy.cc (hunk 5 above)
Comment (RomanDzhabarov) [reply source/common/mongo/proxy.cc]: adding handling of this situation.

File: source/common/mongo/proxy.h (hunk 4 above)
Comment (ccaraman) [reply source/common/mongo/proxy.h]: I would argue that this is more code. There is always a chance of a config being wrong but at least you'll be using the same code that production is using.

Comment (RomanDzhabarov) [reply source/common/mongo/proxy.h]: it's more code when there is an interface and two implementations

File: docs/configuration/network_filters/mongo_proxy_filter.rst [hunk 7]
```diff
@@ -25,6 +26,30 @@ access_log
   path is specified no access logs will be written. Note that access log is also gated by
   :ref:`runtime <config_network_filters_mongo_proxy_runtime>`.
 
+fault
+  *(optional, object)* If specified, the filter will inject faults based on the values in the object.
+
+Fault configuration
+-------------------  
+
+Configuration for fixed duration delays into `Query` mongo operation.
```
Comment (ccaraman) [inline docs/configuration/network_filters/mongo_proxy_filter.rst]: About this configuration, what will the other delays ones look like? If we are to allow per operation configuration, should the name of query be in here.

File: source/common/mongo/proxy.cc [hunk 8]
```diff
@@ -241,5 +245,54 @@ DecoderPtr ProdProxyFilter::createDecoder(DecoderCallbacks& callbacks) {
...
+
+  if (!fault_config_)
+    return result;
+
+  if (!runtime_.snapshot().featureEnabled("mongo.fault.delay.percent",
+                                          fault_config_->delayPercent())) {
+    return result;
+  }
+
+  uint64_t duration = runtime_.snapshot().getInteger("mongo.fault.delay.duration_ms",
```
Comment (ccaraman) [inline source/common/mongo/proxy.cc]: similar to fault filter these keys should be static strings ex: https://github.com/lyft/envoy/blob/master/source/common/http/filter/fault_filter.cc#L24

File: source/common/mongo/proxy.cc [hunk 9]
```diff
@@ -241,5 +245,54 @@ DecoderPtr ProdProxyFilter::createDecoder(DecoderCallbacks& callbacks) {
   return DecoderPtr{new DecoderImpl(callbacks)};
 }
 
+FaultConfig::FaultConfig(const Json::Object& fault_config)
```
Comment (ccaraman) [inline source/common/mongo/proxy.cc]: move this to the header

File: test/common/mongo/proxy_test.cc [hunk 10]
```diff
@@ -64,23 +67,98 @@ class MongoProxyFilterTest : public testing::Test {
...
+  void initializeFilter() {
+    filter_.reset(
+        new TestProxyFilter("test.", store_, runtime_, access_log_, fault_config_, dispatcher_));
     filter_->initializeReadFilterCallbacks(read_filter_callbacks_);
     filter_->onNewConnection();
   }
 
+  void setupDelayFault(uint64_t duration_ms, uint32_t delay_percent, bool enable_fault) {
+    const std::string json_fault_config_template = R"EOF(
+    {{
```
Comment (ccaraman) [inline test/common/mongo/proxy_test.cc]: why the double { ?

File: source/common/mongo/proxy.cc (hunk 8 above)
Comment (RomanDzhabarov) [reply source/common/mongo/proxy.cc]: we are somewhat inconsistent with strings for runtime keys across the envoy code base, no objection to make that const.

File: test/common/mongo/proxy_test.cc (hunk 10 above)
Comment (RomanDzhabarov) [reply test/common/mongo/proxy_test.cc]: this is to make fmt::format happy :) (it's happy now)

File: source/common/mongo/proxy.cc (hunk 5 above)
Comment (RomanDzhabarov) [reply source/common/mongo/proxy.cc]: this is fixed in the next iteration.

File: source/common/mongo/proxy.cc [hunk 11]
```diff
@@ -226,9 +235,13 @@ void ProxyFilter::onEvent(Network::ConnectionEvent event) {
 }
 
 Network::FilterStatus ProxyFilter::onData(Buffer::Instance& data) {
+  if (delay_timer_)
```
Comment (ccaraman) [inline source/common/mongo/proxy.cc]: if we have an active delay here, won't all requests be blocked? ie non query requests are being delayed.

File: source/common/mongo/proxy.cc (hunk 5 above)
Comment (RomanDzhabarov) [reply source/common/mongo/proxy.cc]: this needs to be fixed to capture stats via decode. I'll fix it today.

File: source/common/mongo/proxy.cc (hunk 11 above)
Comment (RomanDzhabarov) [reply source/common/mongo/proxy.cc]: I think we could not do on a per operation delays until we make mongo proxy filter full L7 filter, and not the bump in the wire + tcp proxy.

There is still a problem with capturing stats in here, i'll fix.

**Comment by RomanDzhabarov** [comment]
@ccaraman @mattklein123 one more pass on this?

File: docs/configuration/network_filters/mongo_proxy_filter.rst [hunk 12]
```diff
@@ -151,6 +178,12 @@ mongo.logging_enabled
   % of messages that will be logged. Defaults to 100. If less than 100, queries may be logged
   without replies, etc.
 
+mongo.fault.delay.percent
```
Comment (ccaraman) [inline docs/configuration/network_filters/mongo_proxy_filter.rst]: update keys to match config please

File: source/common/mongo/proxy.cc [hunk 13]
```diff
@@ -50,12 +56,16 @@ ProxyFilter::ProxyFilter(const std::string& stat_prefix, Stats::Scope& scope,
 ProxyFilter::~ProxyFilter() {}
```
Comment (ccaraman) [inline source/common/mongo/proxy.cc]: If there is a disconnect, the timer will still exist and will eventually callback when the cnx doesn't exist.

File: source/common/mongo/proxy.cc [hunk 14]
```diff
@@ -241,5 +254,50 @@ DecoderPtr ProdProxyFilter::createDecoder(DecoderCallbacks& callbacks) {
   return DecoderPtr{new DecoderImpl(callbacks)};
 }
 
+Optional<uint64_t> ProxyFilter::delayDuration() {
+  Optional<uint64_t> result;
+
+  if (!fault_config_)
+    return result;
```
Comment (ccaraman) [inline source/common/mongo/proxy.cc]: please add {} around this if statement

File: source/common/mongo/proxy.cc (hunk 13 above)
Comment (RomanDzhabarov) [reply source/common/mongo/proxy.cc]: when ProxyFilter is destroyed, unique ptr (https://github.com/lyft/envoy/pull/1519/files#diff-8fff5f501b19d4f27e10798d8900a387R182) which holds the timer will deallocate memory used for it.

Comment (mattklein123) [reply source/common/mongo/proxy.cc]: See all other filters. Do to deferred deletion you must destroy timer at disconnect time.

Comment (ccaraman) [reply source/common/mongo/proxy.cc]: Add assert for the delay_timer ex https://github.com/lyft/envoy/blob/master/source/common/http/filter/fault_filter.cc#L76

File: source/common/mongo/proxy.cc [hunk 15]
```diff
@@ -50,12 +56,16 @@ ProxyFilter::ProxyFilter(const std::string& stat_prefix, Stats::Scope& scope,
 ProxyFilter::~ProxyFilter() {}
 
 void ProxyFilter::decodeGetMore(GetMoreMessagePtr&& message) {
```
Comment (ccaraman) [inline source/common/mongo/proxy.cc]: should there be a delay on KillCursors?

File: test/common/mongo/proxy_test.cc [hunk 16]
```diff
@@ -64,23 +67,122 @@ class MongoProxyFilterTest : public testing::Test {
...
   AccessLogSharedPtr access_log_;
+  FaultConfigSharedPtr fault_config_;
   std::unique_ptr<TestProxyFilter> filter_;
   NiceMock<Network::MockReadFilterCallbacks> read_filter_callbacks_;
   Envoy::AccessLog::MockAccessLogManager log_manager_;
 };
 
+TEST_F(MongoProxyFilterTest, DelayFaults) {
+  const uint64_t delay = 10;
+  setupDelayFault(delay, 50, true);
```
Comment (ccaraman) [inline test/common/mongo/proxy_test.cc]: All of your tests are using the same delay of 10, 50 just have on config that you reuse.

File: source/common/mongo/proxy.cc (hunk 13 above)
Comment (RomanDzhabarov) [reply source/common/mongo/proxy.cc]: sounds good

File: source/common/mongo/proxy.cc (hunk 15 above)
Comment (RomanDzhabarov) [reply source/common/mongo/proxy.cc]: as synced offline, let's add this.

File: source/common/mongo/proxy.cc [hunk 17]
```diff
@@ -241,5 +261,51 @@ DecoderPtr ProdProxyFilter::createDecoder(DecoderCallbacks& callbacks) {
...
+  // Delay only if the duration is > 0ms.
+  if (duration > 0) {
+    result.value(duration);
+  }
+
+  return result;
+}
+
+void ProxyFilter::delayInjectionTimerCallback() {
+  delay_timer_.reset();
```
Comment (ccaraman) [inline source/common/mongo/proxy.cc]: Need to disable the timer.  Could make it a helper method.

Comment (RomanDzhabarov) [reply source/common/mongo/proxy.cc]: not necessary, at this point timer triggered already, so no need to disable.

File: test/common/mongo/proxy_test.cc (hunk 16 above)
Comment (RomanDzhabarov) [reply test/common/mongo/proxy_test.cc]: fixed in the next iteration

File: source/common/mongo/proxy.cc (hunk 17 above)
Comment (RomanDzhabarov) [reply source/common/mongo/proxy.cc]: posted.

File: test/common/mongo/proxy_test.cc (hunk 16 above)
Comment (RomanDzhabarov) [reply test/common/mongo/proxy_test.cc]: posted.

File: source/common/json/config_schemas.cc [hunk 18]
```diff
@@ -354,11 +354,34 @@ const std::string Json::Schema::HTTP_CONN_NETWORK_FILTER_SCHEMA(R"EOF(
...
+            "type" : "object",
+            "properties" : {
+              "percent" : {
+                "type" : "integer",
+                "minimum" : 0,
+                "maximum" : 100
+              },
+              "duration_ms" : {
+                "type" : "integer",
+                "minimum" : 0
```
Comment (ccaraman) [inline source/common/json/config_schemas.cc]: add `"exclusiveMinimum" : true` like the fault filter

Comment (RomanDzhabarov) [reply source/common/json/config_schemas.cc]: addressing will post in the next iteration.

File: docs/configuration/network_filters/mongo_proxy_filter.rst [hunk 19]
```diff
@@ -25,6 +26,31 @@ access_log
   path is specified no access logs will be written. Note that access log is also gated by
   :ref:`runtime <config_network_filters_mongo_proxy_runtime>`.
 
+fault
+  *(optional, object)* If specified, the filter will inject faults based on the values in the object.
+
+Fault configuration
+-------------------  
+
+Configuration for MongoDB fixed duration delays. Once delay is injected all of the multiplexed
```
Comment (ccaraman) [inline docs/configuration/network_filters/mongo_proxy_filter.rst]: What is unclear about this text is that I assume the Mongo Filter will work very similar to the HTTP Fault Filter; however, after talking about it in person that isn't the case. 

Please elaborate on the first set of data that arrives with no active timer enabled will get the full delay duration and all subsequent data arriving on that connection up until the timer is 'complete' will be apart of that delay.

File: docs/configuration/network_filters/mongo_proxy_filter.rst [hunk 20]
```diff
@@ -151,6 +178,12 @@ mongo.logging_enabled
   % of messages that will be logged. Defaults to 100. If less than 100, queries may be logged
   without replies, etc.
 
+mongo.fault.fixed_delay.percent
+  % of requests that will be delayed. Defaults to the *percent* specified in the config.
```
Comment (ccaraman) [inline docs/configuration/network_filters/mongo_proxy_filter.rst]: % of requests is incorrect. This would be the number of 'first' incoming data that is received and the delayenabled function(name not sure is correct) says a delay should be injected.

File: docs/configuration/network_filters/mongo_proxy_filter.rst (hunk 19 above)
Comment (ccaraman) [reply docs/configuration/network_filters/mongo_proxy_filter.rst]: Also, do we have an issue tracking making this fault filter 'smarter' and to rework it so either a consistent delay is applied every new call of OnData or per operation?

File: source/common/mongo/proxy.cc [hunk 21]
```diff
@@ -36,38 +42,46 @@ void AccessLog::logMessage(const Message& message, bool full,
...
     // If we are not logging at the connection level, just release the shared pointer so that we
     // don't ever log.
     access_log_.reset();
   }
 }
 
-ProxyFilter::~ProxyFilter() {}
+ProxyFilter::~ProxyFilter() { ASSERT(!delay_timer_); }
 
 void ProxyFilter::decodeGetMore(GetMoreMessagePtr&& message) {
```
Comment (ccaraman) [inline source/common/mongo/proxy.cc]: what about decodeReply?

Comment (ccaraman) [reply source/common/mongo/proxy.cc]: Is there a better place to add the cal to tryInjectDelay so we don't have too start every method with `tryInjectDelay()`

File: docs/configuration/network_filters/mongo_proxy_filter.rst [hunk 22]
```diff
@@ -25,6 +26,31 @@ access_log
...
   :ref:`runtime <config_network_filters_mongo_proxy_runtime>`.
 
+fault
+  *(optional, object)* If specified, the filter will inject faults based on the values in the object.
+
+Fault configuration
+-------------------  
+
+Configuration for MongoDB fixed duration delays. Delays are applied to the following MongoDB operations: Query, Insert,
+GetMore, KillCursors. First command that arrives with no active delays enabled will get the full delay duration and all subsequent data arriving on that connection up until the timer is 'complete' will be apart of that delay.
```
Comment (ccaraman) [inline docs/configuration/network_filters/mongo_proxy_filter.rst]: over character limit for the line

File: source/common/mongo/proxy.cc (hunk 21 above)
Comment (RomanDzhabarov) [reply source/common/mongo/proxy.cc]: no need for Reply as it's the response for the original injected request (query, insert, etc)

Comment (RomanDzhabarov) [reply source/common/mongo/proxy.cc]: let's see for the tryInjectDelay() place

File: docs/configuration/network_filters/mongo_proxy_filter.rst (hunk 22 above)
Comment (ccaraman) [reply docs/configuration/network_filters/mongo_proxy_filter.rst]: Can you find the more appropriate term for timer is complete? Triggered?

File: test/common/mongo/proxy_test.cc [hunk 23]
```diff
@@ -64,23 +67,127 @@ class MongoProxyFilterTest : public testing::Test {
...
   std::shared_ptr<Filesystem::MockFile> file_{new NiceMock<Filesystem::MockFile>()};
   AccessLogSharedPtr access_log_;
+  FaultConfigSharedPtr fault_config_;
   std::unique_ptr<TestProxyFilter> filter_;
   NiceMock<Network::MockReadFilterCallbacks> read_filter_callbacks_;
   Envoy::AccessLog::MockAccessLogManager log_manager_;
 };
 
+TEST_F(MongoProxyFilterTest, DelayFaults) {
+  const uint64_t delay = 10;
```
Comment (ccaraman) [inline test/common/mongo/proxy_test.cc]: del this line

File: test/common/mongo/proxy_test.cc [hunk 24]
```diff
@@ -320,6 +442,14 @@ TEST_F(MongoProxyFilterTest, EmptyActiveQueryList) {
 }
 
 TEST_F(MongoProxyFilterTest, ConnectionDestroyLocal) {
+  const uint64_t delay = 10;
```
Comment (ccaraman) [inline test/common/mongo/proxy_test.cc]: delete this line

File: test/common/mongo/proxy_test.cc [hunk 25]
```diff
@@ -330,12 +460,21 @@ TEST_F(MongoProxyFilterTest, ConnectionDestroyLocal) {
...
   filter_->onData(fake_data_);
 
+  EXPECT_CALL(*delay_timer, disableTimer());
   read_filter_callbacks_.connection_.raiseEvent(Network::ConnectionEvent::RemoteClose);
   EXPECT_EQ(1U, store_.counter("test.cx_destroy_local_with_active_rq").value());
   EXPECT_EQ(0U, store_.counter("test.cx_destroy_remote_with_active_rq").value());
 }
 
 TEST_F(MongoProxyFilterTest, ConnectionDestroyRemote) {
+  const uint64_t delay = 10;
```
Comment (ccaraman) [inline test/common/mongo/proxy_test.cc]: delete this line

File: test/server/config/network/mongo_proxy_test.cc [hunk 26]
```diff
@@ -0,0 +1,224 @@
...
+        "duration_ms": -1
+      }
+    }
+  }
+  )EOF";
+
+  handleInvalidConfiguration(json_string);
+}
+
+TEST(MongoFilterConfigTest, InvalidFaultsDelayPercent) {
```
Comment (ccaraman) [inline test/server/config/network/mongo_proxy_test.cc]: for future reference, there isn't a need to test all permutations of wrong configs.

Comment (RomanDzhabarov) [reply test/server/config/network/mongo_proxy_test.cc]: haha, full coverage

File: source/common/mongo/proxy.h [hunk 27]
```diff
@@ -8,6 +8,8 @@
 
 #include "envoy/access_log/access_log.h"
 #include "envoy/common/time.h"
+#include "envoy/event/dispatcher.h"
```
Comment (ccaraman) [inline source/common/mongo/proxy.h]: is this still needed?

File: docs/configuration/network_filters/mongo_proxy_filter.rst (hunk 19 above)
Comment (ccaraman) [reply docs/configuration/network_filters/mongo_proxy_filter.rst]: @RomanDzhabarov just in case you didn't see this comment.

File: docs/configuration/network_filters/mongo_proxy_filter.rst [hunk 28]
```diff
@@ -151,6 +179,14 @@ mongo.logging_enabled
   % of messages that will be logged. Defaults to 100. If less than 100, queries may be logged
   without replies, etc.
 
+mongo.fault.fixed_delay.percent
+  Probability of an elibible MongoDB operation to be affected by
```
Comment (ccaraman) [inline docs/configuration/network_filters/mongo_proxy_filter.rst]: should there be a note if there is no entry in the config.

File: source/common/mongo/proxy.h (hunk 27 above)
Comment (RomanDzhabarov) [reply source/common/mongo/proxy.h]: nope

File: docs/configuration/network_filters/mongo_proxy_filter.rst (hunk 28 above)
Comment (RomanDzhabarov) [reply docs/configuration/network_filters/mongo_proxy_filter.rst]: `percent` is a required field, should be in the config.

File: docs/configuration/network_filters/mongo_proxy_filter.rst [hunk 29]
```diff
@@ -25,6 +26,32 @@ access_log
...
   :ref:`runtime <config_network_filters_mongo_proxy_runtime>`.
 
+fault
+  *(optional, object)* If specified, the filter will inject faults based on the values in the object.
+
+Fault configuration
+-------------------  
+
+Configuration for MongoDB fixed duration delays. Delays are applied to the following MongoDB operations: Query, Insert,
+GetMore, KillCursors. First command that arrives with no active delays enabled will get the full delay duration and
```
Comment (ccaraman) [inline docs/configuration/network_filters/mongo_proxy_filter.rst]: The second sentence isn't very clear. It could be
`Once an active delay is in progress, all incoming data up until the timer event fires will be apart of the delay.`

Comment (ccaraman) [reply docs/configuration/network_filters/mongo_proxy_filter.rst]: And expand that the stat only increments on the delay event being started not per operation.

File: docs/configuration/network_filters/mongo_proxy_filter.rst (hunk 28 above)
Comment (ccaraman) [reply docs/configuration/network_filters/mongo_proxy_filter.rst]: nope, not if fault object isn't there

Comment (RomanDzhabarov) [reply docs/configuration/network_filters/mongo_proxy_filter.rst]: This runtime setting is taken into consideration only if `fault` object is specified for mongo_proxy filter.

I can put into the comments that this runtime setting is used only if `fault` configuration is defined for the proxy.

File: docs/configuration/network_filters/mongo_proxy_filter.rst (hunk 19 above)
Comment (RomanDzhabarov) [reply docs/configuration/network_filters/mongo_proxy_filter.rst]: yup, https://github.com/lyft/envoy/issues/530. I'll link to the ticket and explain what's missing.

File: docs/configuration/network_filters/mongo_proxy_filter.rst (hunk 29 above)
Comment (RomanDzhabarov) [reply docs/configuration/network_filters/mongo_proxy_filter.rst]: yes, will explain more on stats.

Comment (RomanDzhabarov) [reply docs/configuration/network_filters/mongo_proxy_filter.rst]: should be addressed in the last iteration.

File: source/common/mongo/proxy.h (hunk 27 above)
Comment (RomanDzhabarov) [reply source/common/mongo/proxy.h]: still needed, but I moved that to the `cc` file.

File: docs/configuration/network_filters/mongo_proxy_filter.rst [hunk 30]
```diff
@@ -25,6 +26,32 @@ access_log
...
   :ref:`runtime <config_network_filters_mongo_proxy_runtime>`.
 
+fault
+  *(optional, object)* If specified, the filter will inject faults based on the values in the object.
+
+Fault configuration
+-------------------  
+
+Configuration for MongoDB fixed duration delays. Delays are applied to the following MongoDB operations: Query, Insert,
+GetMore, KillCursors. Once an active delay is in progress, all incoming data up until the timer event fires
```
Comment (mattklein123) [inline docs/configuration/network_filters/mongo_proxy_filter.rst]: "and KillCursors"

File: docs/configuration/network_filters/mongo_proxy_filter.rst [hunk 31]
```diff
@@ -25,6 +26,32 @@ access_log
...
+will be apart of the delay.
```
Comment (mattklein123) [inline docs/configuration/network_filters/mongo_proxy_filter.rst]: "a part"

File: docs/configuration/network_filters/mongo_proxy_filter.rst [hunk 32]
```diff
@@ -25,6 +26,32 @@ access_log
...
+
+  {
+    "fixed_delay": {
+      "percent": "...",
+      "duration_ms": "..."
+    }
+  }
+
+percent
+  *(required, integer)* Probability of an elibible MongoDB operation to be affected by injected
```
Comment (mattklein123) [inline docs/configuration/network_filters/mongo_proxy_filter.rst]: typo elibible

File: docs/configuration/network_filters/mongo_proxy_filter.rst [hunk 33]
```diff
@@ -25,6 +26,32 @@ access_log
...
+  faults when there is no active faults. Valid values are integers in a range of [0, 100].
```
Comment (mattklein123) [inline docs/configuration/network_filters/mongo_proxy_filter.rst]: "are no active faults"

File: docs/configuration/network_filters/mongo_proxy_filter.rst [hunk 34]
```diff
@@ -25,6 +26,32 @@ access_log
...
+
+duration_ms
+  *(required, integer)* Non negative delay duration in milliseconds.
```
Comment (mattklein123) [inline docs/configuration/network_filters/mongo_proxy_filter.rst]: "Non-negative"

File: docs/configuration/network_filters/mongo_proxy_filter.rst [hunk 35]
```diff
@@ -38,6 +65,7 @@ following statistics:
   :widths: 1, 1, 2
 
   decoding_error, Counter, Number of MongoDB protocol decoding errors
+  delay_injected, Counter, Number of times delay is started and it's not number of operations delayed
```
Comment (mattklein123) [inline docs/configuration/network_filters/mongo_proxy_filter.rst]: grammar / can't parse

File: docs/configuration/network_filters/mongo_proxy_filter.rst (hunk 28 above)
Comment (mattklein123) [reply docs/configuration/network_filters/mongo_proxy_filter.rst]: typo elibible, "are no active faults"

File: include/envoy/network/filter.h [hunk 36]
```diff
@@ -77,7 +77,7 @@ class ReadFilter {
   virtual ~ReadFilter() {}
 
   /**
-   * Called when data is read on the connection.
+   * Called when data is ready on the connection.
```
Comment (mattklein123) [inline include/envoy/network/filter.h]: ? I would live old comment.

File: source/common/mongo/proxy.cc [hunk 37]
```diff
@@ -17,6 +18,12 @@
 namespace Envoy {
 namespace Mongo {
 
+const std::string ProxyFilter::FIXED_DELAY_PERCENT_KEY = "mongo.fault.fixed_delay.percent";
```
Comment (mattklein123) [inline source/common/mongo/proxy.cc]: All of these should be part of a constant singleton class like Http::Headers

File: source/common/mongo/proxy.cc [hunk 38]
```diff
@@ -211,24 +226,32 @@ void ProxyFilter::doDecode(Buffer::Instance& buffer) {
...
+      delay_timer_->disableTimer();
+      delay_timer_.reset();
+    }
+  }
+
   if (event == Network::ConnectionEvent::RemoteClose && !active_query_list_.empty()) {
     stats_.cx_destroy_local_with_active_rq_.inc();
-  }
-  if (event == Network::ConnectionEvent::LocalClose && !active_query_list_.empty()) {
+  } else if (event == Network::ConnectionEvent::LocalClose && !active_query_list_.empty()) {
```
Comment (mattklein123) [inline source/common/mongo/proxy.cc]: unnecessary change

File: source/common/mongo/proxy.cc [hunk 39]
```diff
@@ -241,5 +264,51 @@ DecoderPtr ProdProxyFilter::createDecoder(DecoderCallbacks& callbacks) {
...
+
+  if (!fault_config_) {
+    return result;
+  }
+
+  if (!runtime_.snapshot().featureEnabled(FIXED_DELAY_PERCENT_KEY, fault_config_->delayPercent())) {
+    return result;
+  }
+
+  uint64_t duration =
```
Comment (mattklein123) [inline source/common/mongo/proxy.cc]: const

File: source/common/mongo/proxy.cc [hunk 40]
```diff
@@ -241,5 +264,51 @@ DecoderPtr ProdProxyFilter::createDecoder(DecoderCallbacks& callbacks) {
...
+}
+
+void ProxyFilter::tryInjectDelay() {
+  // Do not try to inject delays if there is an active delay.
+  // Make sure to capture stats for the request otherwise.
+  if (delay_timer_) {
+    return;
+  }
+
+  Optional<uint64_t> delay_ms = delayDuration();
```
Comment (mattklein123) [inline source/common/mongo/proxy.cc]: const

File: source/common/mongo/proxy.h [hunk 41]
```diff
@@ -82,7 +106,7 @@ class ProxyFilter : public Network::Filter,
                     Logger::Loggable<Logger::Id::mongo> {
 public:
   ProxyFilter(const std::string& stat_prefix, Stats::Scope& scope, Runtime::Loader& runtime,
-              AccessLogSharedPtr access_log);
+              AccessLogSharedPtr access_log, FaultConfigSharedPtr fault_config);
```
Comment (mattklein123) [inline source/common/mongo/proxy.h]: `const FaultConfigSharedPtr&`

File: source/common/mongo/proxy.h [hunk 42]
```diff
@@ -73,6 +76,27 @@ class AccessLog {
...
+
+  uint32_t delayPercent() const { return delay_percent_; }
+  uint64_t delayDuration() const { return duration_ms_; }
+
+private:
+  const uint32_t delay_percent_;
+  const uint64_t duration_ms_;
+};
+
+typedef std::shared_ptr<FaultConfig> FaultConfigSharedPtr;
```
Comment (mattklein123) [inline source/common/mongo/proxy.h]: `typedef std::shared_ptr<const FaultConfig> FaultConfigSharedPtr;`

File: source/server/config/network/mongo_proxy.cc [hunk 43]
```diff
@@ -24,9 +24,15 @@ MongoProxyFilterConfigFactory::createFilterFactory(const Json::Object& config,
         new Mongo::AccessLog(config.getString("access_log"), context.accessLogManager()));
   }
 
-  return [stat_prefix, &context, access_log](Network::FilterManager& filter_manager) -> void {
-    filter_manager.addFilter(Network::FilterSharedPtr{
-        new Mongo::ProdProxyFilter(stat_prefix, context.scope(), context.runtime(), access_log)});
+  Mongo::FaultConfigSharedPtr fault_config;
+  if (config.hasObject("fault")) {
+    fault_config.reset(new Mongo::FaultConfig(*config.getObject("fault")));
```
Comment (mattklein123) [inline source/server/config/network/mongo_proxy.cc]: std::make_shared

File: source/server/config/network/mongo_proxy.cc [hunk 44]
```diff
@@ -24,9 +24,15 @@ MongoProxyFilterConfigFactory::createFilterFactory(const Json::Object& config,
...
+  }
+
+  return [stat_prefix, &context, access_log,
+          fault_config](Network::FilterManager& filter_manager) -> void {
+    filter_manager.addFilter(Network::FilterSharedPtr{new Mongo::ProdProxyFilter(
```
Comment (mattklein123) [inline source/server/config/network/mongo_proxy.cc]: switch to std::make_shared

File: test/common/mongo/proxy_test.cc [hunk 45]
```diff
@@ -19,6 +19,7 @@
 #include "gtest/gtest.h"
 
 namespace Envoy {
+using testing::AnyNumber;
```
Comment (mattklein123) [inline test/common/mongo/proxy_test.cc]: move usings outside of namespace

**Comment by RomanDzhabarov** [comment]
@mattklein123 addressed comments.

File: source/common/mongo/proxy.cc [hunk 46]
```diff
@@ -36,38 +37,47 @@ void AccessLog::logMessage(const Message& message, bool full,
 }
 
 ProxyFilter::ProxyFilter(const std::string& stat_prefix, Stats::Scope& scope,
-                         Runtime::Loader& runtime, AccessLogSharedPtr access_log)
+                         Runtime::Loader& runtime, AccessLogSharedPtr access_log,
+                         const FaultConfigSharedPtr fault_config)
```
Comment (mattklein123) [inline source/common/mongo/proxy.cc]: `const FaultConfigSharedPtr&`
